allauth/socialaccount/providers/apple/views.py

```diff
@@ -0,0 +1,130 @@
+    authorize_url = "https://appleid.apple.com/auth/authorize"
+    public_key_url = "https://appleid.apple.com/auth/keys"
+
+    def get_public_key(self, id_token):
+        """
+        Get the public key which matches the `kid` in the id_token header.
+        """
+        kid = jwt.get_unverified_header(id_token)["kid"]
+        apple_public_key = [
+            d
+            for d in requests.get(self.public_key_url).json()["keys"]
+            if d["kid"] == kid
+        ][0]
+        public_key = jwt.algorithms.RSAAlgorithm.from_jwk(
+            json.dumps(apple_public_key)
+        )
+        return public_key
+
+    def get_client_id(self, provider):
+        app = SocialApp.objects.get(provider=provider.id)
+        return [aud.strip() for aud in app.client_id.split(",")]
+
+    def parse_token(self, data):
+        try:
+            token = SocialToken(token=data["access_token"])
+            token.token_secret = data.get("refresh_token", "")
+
+            public_key = self.get_public_key(data["id_token"])
+            provider = self.get_provider()
+            allowed_auds = self.get_client_id(provider)
+
+            token.user_data = jwt.decode(
+                data["id_token"],
+                public_key,
+                algorithms=["RS256"],
+                verify=True,
+                audience=allowed_auds,
+            )
+            expires_in = data.get(self.expires_in_key, None)
+            if expires_in:
+                token.expires_at = timezone.now() + timedelta(
+                    seconds=int(expires_in)
+                )
+            return token
+        except jwt.PyJWTError as e:
+            raise OAuth2Error("Invalid id_token") from e
+
+    def complete_login(self, request, app, token, **kwargs):
+        extra_data = token.user_data
+        login = self.get_provider().sociallogin_from_response(
+            request, extra_data
+        )
+        login.state["id_token"] = token.user_data
+        return login
+
+
+class AppleOAuth2ClientMixin:
+    def get_client(self, request, app):
+        client = super().get_client(request, app)
+        apple_client = AppleOAuth2Client(
+            client.request,
+            client.consumer_key,
+            client.consumer_secret,
+            client.access_token_method,
+            client.access_token_url,
+            client.callback_url,
+            client.scope,
+            key=client.key,
+            cert=client.cert,
+        )
+        return apple_client
+
+
+class AppleOAuth2LoginView(AppleOAuth2ClientMixin, OAuth2LoginView):
+    """
+    Custom AppleOAuth2LoginView to return AppleOAuth2Client
+    """
+
+    pass
+
+
+class AppleOAuth2CallbackView(AppleOAuth2ClientMixin, OAuth2CallbackView):
+    """
+    Custom OAuth2CallbackView because `Sign In With Apple`:
+        * returns AppleOAuth2Client
+        * Apple requests callback by POST
+    """
+
+    def dispatch(self, request, *args, **kwargs):
+        if request.method == "POST":
```

Comment:
What is the reason to redirect to the same path with `GET` method? We need to fix the `OAuth2CallbackView` to use `get_request_param` for getting the code.
Change the follwoing
`access_token = client.get_access_token(request.GET['code'])`
to
`access_token = client.get_access_token(get_request_param('code'))`

Reply:
I am not sure to update `OAuth2CallbackView`, that's why. If you are ready to help, please commit changes to my branch 🤷 @safwanrahman

Reply:
@0leg5ergeev Yeah,I will make a PR to your fork.